### ISSUE TITLE
💡 RFC: Add compiler warning when you need to use `<Fragment>`.

### ISSUE THREAD

**Issue by jasikpark** [snowpackjs/astro#1049]
### Background & Motivation

I want better error reporting. (Also is this the right issue type? Maybe there could be a "diagnostics" error message?)

<details>
<summary>
Quote from @brycewray via Discord
</summary>
<quote>
Would explain why this didn't work, then...
<pre><code>
<div>
{page.data.map((post) => (
  <p>The title is {post.title}</p>
))}
</div>
</code></pre>

...which produced:
<pre><code>
[access] /posts/
[getStaticPaths] invalid path param: page. A string value was expected, but got `undefined`.
[8:54 PM]
</pre></code>
And I am using a <code>[...page].astro</code> file
[8:56 PM]
But I will study this link. Thanks very much.

1

BryceWray — Today at 9:15 PM
So, just noting that this works:
<pre><code>
{page.data.map((post => 
  <p>Title: {post.title}</p>
))}
</code></pre>

...but adding another line (and I do have a date item in each post's front matter)...

<pre><code>
{page.data.map((post => 
  <p>Title: {post.title}</p>
  <p>Date: {post.date}</p>
))}
</code></pre>

...causes the error message:

<pre><code>
 /Users/thewrays/astro-site/src/pages/posts/[...page].astro:21:4

 𝘅 Expected ")" but found "h"

     19 | {page.data.map((post => 
     20 |   <p>Title: {post.title}</p>
   > 21 |   <p>Date: {post.date}</p>
        |    ^
     22 | ))}

(node:75632) UnhandledPromiseRejectionWarning: Error: Unable to compile expression
</code></pre>

...with a lot more after this. Anyway, I understand it's brittle right now. Will just wait and see.
</quote>
</details>

### Possible solutions
Currently, we get an `UnhandledPromiseRejectionWarning` but I'd like to see if we could detect this condition of "tries to use multiple JSX-like syntax not at the top level" and add a warning for it!
### Open Questions
How would one add diagnostics to the compiler at this day and age? 🤔 

### Help make it happen!

- [ ] I am willing to submit a PR to implement this change.
- [X] I am willing to submit a PR to implement this change, but would need some guidance.
- [ ] I am not willing to submit a PR to implement this change.

**Comment by drwpow** [comment]
I think this is something we all want! Just a gut reaction, to take with a grain of salt, but I wonder if implementing this and detecting _when_ a fragment needs to be used, could just be used to add the fragment for you behind-the-scenes so you never needed it in the first place.

Some other similar parser issues come to mind that may or may not be related (mentioned because some of these can be solved with fragments):

- #1133
- #1061 
- #1060 
- #1059

**Comment by jasikpark** [comment]
It definitely would be great to make that work!

**Comment by matthewp** [comment]
From our most recent demo, @natemoo-re showed that using Fragment is not at all needed in the new compiler. Happy to keep the issue open if you want until that's confirmed and out.

**Comment by brycewray** [comment]
@matthewp Thanks. Currently, I'm hung up on a totally different one (https://github.com/snowpackjs/astro/issues/1098) — but am glad to hear about this, as well!

**Comment by natemoo-re** [comment]
Confirming that this is fixed in the `astro@next` preview!

Will put this in our milestone to be closed when that release is stable.

**Comment by natemoo-re** [comment]
Following up on my previous comment, the team has decided to close out issues that have been confirmed as fixed by [astro@0.21.0-next.0](https://github.com/snowpackjs/astro/releases/tag/astro%400.21.0-next.0), [astro@0.21.0-next.1](https://github.com/snowpackjs/astro/releases/tag/astro%400.21.0-next.1), or [astro@0.21.0-next.2](https://github.com/snowpackjs/astro/releases/tag/astro%400.21.0-next.2). Our hope is that this will help the [v0.21 milestone](https://github.com/snowpackjs/astro/milestone/1) remain as actionable as possible.

To verify that this issue has been fixed, you may
- Open a new [`astro@next` project on StackBlitz](https://astro.new/minimal@next)
- Locally upgrade to the `astro@next` preview
    ```shell
    npm i astro@next
    # If using Framework renderers...
    npm i @astrojs/renderer-name@next
    ```

If you think we made a mistake, **please reopen this issue**!

If you run into any other problems with `astro@next`, please [open a new issue](https://github.com/snowpackjs/astro/issues/new?assignees=&labels=0.21&template=---01-bug-report.yml&title=%F0%9F%90%9B+BUG%3Aastro@next).
